pallets/asset-index/src/lib.rs

```diff
@@ -673,6 +713,51 @@ pub mod pallet {
 			share.checked_div(&nav).ok_or_else(|| ArithmeticError::Overflow.into())
 		}
 
+		/// Consolidate and clean deposits while completing withdraw
+		fn do_consolidate_deposits(caller: &T::AccountId, mut amount: T::Balance) -> Result<T::Balance, DispatchError> {
+			let withdrawal_amount = amount.clone();
+			<Deposits<T>>::try_mutate_exists(&caller, |maybe_depositing| -> Result<T::Balance, DispatchError> {
+				let depositing = maybe_depositing.take().ok_or(<Error<T>>::NoDeposits)?;
+				let mut start: T::BlockNumber = 0_u32.into();
+
+				let still_depositing: Vec<_> = depositing
+					.into_iter()
+					.filter_map(|(mut index_tokens, block_number)| {
+						// check depositing duration
+						if block_number < start || start == 0_u32.into() {
+							start = block_number;
+						}
+
+						// consolidate deposits
+						if block_number > frame_system::Pallet::<T>::block_number() {
+							Some((index_tokens, block_number))
+						} else if amount >= index_tokens {
+							amount = amount.saturating_sub(index_tokens);
+							None
+						} else {
+							index_tokens = index_tokens.saturating_sub(amount);
+							amount = Default::default();
+							Some((index_tokens, block_number))
+						}
+					})
+					.collect();
+
+				if !still_depositing.is_empty() {
+					// still have redemptions pending
+					*maybe_depositing = Some(still_depositing.try_into().map_err(|_| <Error<T>>::InsufficientDeposit)?);
+				}
+
+				if amount != Default::default() {
+					return Err(<Error<T>>::InsufficientDeposit.into());
+				}
+
+				Ok(T::RedemptionFee::redemption_fee(
+					frame_system::Pallet::<T>::block_number().saturating_sub(start),
+					withdrawal_amount,
+				))
```

Comment:
we need to call this for every entry in `deposits`, so execute this in the loop, because the fee is dependent on how long the deposit spent in the index. And then we need to summarize the total fee.